src/lib/rpc/bestRpcTracker.ts

```diff
@@ -0,0 +1,346 @@
+
+// DataStore field used for probing
+const PROBE_SAMPLE_FIELD = "minCollateralFactor";
+// Markets used for `PROBE_SAMPLE_FIELD` reading
+const PROBE_SAMPLE_MARKET = {
+  [ARBITRUM]: "0x70d95587d40A2caf56bd97485aB3Eec10Bee6336", // ETH/USD
+  [AVALANCHE]: "0xB7e69749E3d2EDd90ea59A4932EFEa2D41E245d7", // ETH/USD
+  [AVALANCHE_FUJI]: "0xbf338a6C595f06B7Cfff2FA8c958d49201466374", // ETH/USD
+};
+
+type ProbeData = {
+  url: string;
+  isSuccess: boolean;
+  responseTime: number;
+  blockNumber: number | null;
+  timestamp: Date;
+};
+
+type ProviderData = {
+  url: string;
+  provider: Provider;
+};
+
+type RpcTrackerState = {
+  [chainId: number]: {
+    lastUsage: Date | null;
+    currentBestProviderUrl: string;
+    providers: {
+      [providerUrl: string]: ProviderData;
+    };
+  };
+};
+
+let trackingIntervalId: number | null = null;
+let trackerState: RpcTrackerState | null = null;
+
+function initRpcTracking() {
+  trackerState = initTrackerState();
+
+  if (trackingIntervalId) {
+    clearInterval(trackingIntervalId);
+  }
+  measureRpcData({ warmUp: true });
+  trackingIntervalId = window.setInterval(() => measureRpcData(), PROBE_INTERVAL);
+}
+
+function measureRpcData({ warmUp = false } = {}) {
+  if (!trackerState) {
+    throw new Error("RPC tracker state is not initialized");
+  }
+
+  entries(trackerState).forEach(async ([chainIdRaw, chainTracker]) => {
+    const chainId = Number(chainIdRaw);
+    const providers = Object.values(chainTracker.providers);
+
+    const isUnusedChain =
+      !chainTracker.lastUsage ||
+      differenceInMilliseconds(Date.now(), chainTracker.lastUsage) > DISABLE_UNUSED_TRACKING_TIMEOUT;
+    const isTrackingEnabled = (warmUp || !isUnusedChain) && providers.length > 1;
+
+    if (!isTrackingEnabled) {
+      return;
+    }
+
+    const probePromises = providers.map((providerInfo) => {
+      return probeRpc(chainId, providerInfo.provider, providerInfo.url);
+    });
+
+    const probeResults = await Promise.all(probePromises);
+    const successProbeResults = probeResults.filter((probe) => probe.isSuccess);
+
+    if (!successProbeResults.length) {
+      setCurrentProvider(chainId, getFallbackRpcUrl(chainId));
+
+      return;
+    }
+
+    const probeResultsByBlockNumber = orderBy(successProbeResults, ["blockNumber"], ["desc"]);
+    let bestBlockNumberProbe = probeResultsByBlockNumber[0];
+
+    const probeStats = probeResultsByBlockNumber.map((probe, i, arr) => {
+      let isValid = probe.isSuccess;
+
+      const bestBlockNumber = bestBlockNumberProbe.blockNumber;
+      const currProbeBlockNumber = probe.blockNumber;
+      const nextProbeBlockNumber = arr[i + 1]?.blockNumber;
+
+      // Rare case when RPC returned a block number from the far future
+      if (
+        i === 0 &&
+        currProbeBlockNumber &&
+        nextProbeBlockNumber &&
+        currProbeBlockNumber - nextProbeBlockNumber > BLOCK_FROM_FUTURE_THRESHOLD
+      ) {
```

Comment:
this validation doesn't handle a case when there are 3 probes:
1. block = 1001
2. block = null
3. block = 1

since only (1) and (2) are compared to each other, (1) will be considered valid